File: python/day7.py
import itertools

with open("input/day7.txt", "r") as f:
    lines = f.readlines()
    bag_container = {}
    n = 0
    for line in lines:
        parsed_bags = (
            line.strip()
            .replace(" bags.", "")
            .replace("bag.", "")
            .replace("bags", "")
            .replace("bag", "")
            .split(" contain ")
        )
        other_bags = parsed_bags[1:]
        this_bag = parsed_bags[0].strip()
        bag_to_bag_n = {}
        for other_bag in other_bags[0].split(", "):
            if other_bag == "no other":
                continue
            n_other = int(other_bag[0])
            name_other = other_bag[1:].strip()
            bag_to_bag_n[name_other] = n_other
        bag_container[this_bag.strip()] = bag_to_bag_n


def colors_contained_by_one_color(this_color) -> set[str]:
    if this_color not in bag_container:
        return set()
    contained_bags = bag_container[this_color]
    return set(contained_bags) | set(
        itertools.chain(*[colors_contained_by_one_color(x) for x in contained_bags])
    )


def bags_inside_one_bag(color) -> int:
    if color not in bag_container:
        return 0
    else:
        contained_bags = sum(bag_container[color].values())

        return contained_bags + sum(
            [
                n_contained * bags_inside_one_bag(contained_color)
                for contained_color, n_contained in bag_container[color].items()
            ]
        )


all_colors = bag_container.keys()
all_bags_colors = map(
    lambda color: (color, colors_contained_by_one_color(color)), all_colors
)
colors_with_gold = filter(
    lambda color_container: "shiny gold" in color_container[1], all_bags_colors
)

res = len(list(colors_with_gold))
print(res)
print(bags_inside_one_bag("shiny gold"))

File: src/bin/day7.rs
use std::{collections::HashMap, time::Instant};
const GOLD: &str = "shiny gold";
fn does_bag_contain_shiny<'a>(
    bag_color: &'a str,
    bag_to_bag_to_n: &HashMap<&str, HashMap<&'a str, u32>>,
    cache: &mut HashMap<&'a str, bool>,
) -> bool {
    // This memoisation makes it around 10x faster
    if let Some(res) = cache.get(bag_color) {
        return *res;
    }
    let res = match bag_to_bag_to_n.get(bag_color) {
        None => false,
        Some(contained_bags_to_n) if contained_bags_to_n.contains_key(GOLD) => true,
        Some(contained_bags_to_n) => contained_bags_to_n
            .keys()
            .map(|k| does_bag_contain_shiny(k, bag_to_bag_to_n, cache))
            .any(|v| v == true),
    };
    cache.insert(bag_color, res);
    res
}

fn count_bags_inside_bag(color: &str, bag_to_bag_to_n: &HashMap<&str, HashMap<&str, u32>>) -> u32 {
    match bag_to_bag_to_n.get(color) {
        None => 0,
        Some(contained_bags) => {
            let n_bags_inside_this: u32 = contained_bags.values().sum();
            let bags_inside_bags: u32 = contained_bags
                .iter()
                .map(|(k, v)| v * count_bags_inside_bag(k, bag_to_bag_to_n))
                .sum();
            n_bags_inside_this + bags_inside_bags
        }
    }
}
fn main() {
    let start = Instant::now();
    let input = include_str!("../../input/day7.txt");
    let clean_input = input
        .replace(".", "")
        .replace(" bags", "")
        .replace(" bag", "");
    let mut bag_to_bag_to_n: HashMap<&str, HashMap<&str, u32>> = HashMap::new();
    clean_input.lines().for_each(|lin| {
        let mut two_parts = lin.split(" contain ");
        let bag_name = two_parts.next().unwrap();
        let mut bag_to_n: HashMap<&str, u32> = HashMap::new();
        two_parts.next().unwrap().split(", ").for_each(|other_bag| {
            if other_bag != "no other" {
                let n = other_bag.chars().nth(0).unwrap().to_digit(10).unwrap();
                let inner_bag_name: &str = &other_bag[2..];
                bag_to_n.insert(inner_bag_name, n);
            }
        });
        if !bag_to_n.is_empty() {
            bag_to_bag_to_n.insert(bag_name, bag_to_n);
        }
    });
    // Part 1
    let mut cache = HashMap::new();
    let n_contain_shiny = bag_to_bag_to_n
        .keys()
        .map(|k| does_bag_contain_shiny(k, &bag_to_bag_to_n, &mut cache))
        .filter(|v| *v == true)
        .count();
    println!("{}", n_contain_shiny);
    // Part 2
    let counted_bags = count_bags_inside_bag(GOLD, &bag_to_bag_to_n);
    println!("{}", counted_bags);
    let duration = start.elapsed();

    println!("Time elapsed in expensive_function() is: {:?}", duration);
}
